src/Shared/runtime/Http3/QPack/QPackDecoder.cs

```diff
@@ -105,64 +103,53 @@ private enum State
         private const int StringLengthPrefix = 7;
         private const byte HuffmanMask = 0x80;
 
-        private const int DefaultStringBufferSize = 64;
+        private const int HeaderStaticIndexUnset = -1; // Static index starts at 0
 
         private readonly int _maxHeadersLength;
         private State _state = State.RequiredInsertCount;
 
-        private byte[] _stringOctets;
-        private byte[] _headerNameOctets;
-        private byte[] _headerValueOctets;
-
         // s is used for whatever s is in each field. This has multiple definition
         private bool _huffman;
-        private int? _index;
 
         private byte[]? _headerName;
+        private int _headerStaticIndex;
         private int _headerNameLength;
         private int _headerValueLength;
         private int _stringLength;
         private int _stringIndex;
+
         private IntegerDecoder _integerDecoder;
+        private byte[]? _stringOctets;
+        private byte[]? _headerNameOctets;
+        private byte[]? _headerValueOctets;
+        private (int start, int length)? _headerNameRange;
+        private (int start, int length)? _headerValueRange;
 
         private static ArrayPool<byte> Pool => ArrayPool<byte>.Shared;
 
-        private static void ReturnAndGetNewPooledArray(ref byte[] buffer, int newSize)
-        {
-            byte[] old = buffer;
-            buffer = null!;
-
-            Pool.Return(old, clearArray: true);
-            buffer = Pool.Rent(newSize);
-        }
-
         public QPackDecoder(int maxHeadersLength)
         {
             _maxHeadersLength = maxHeadersLength;
-
-            // TODO: make allocation lazy? with static entries it's possible no buffers will be needed.
-            _stringOctets = Pool.Rent(DefaultStringBufferSize);
-            _headerNameOctets = Pool.Rent(DefaultStringBufferSize);
-            _headerValueOctets = Pool.Rent(DefaultStringBufferSize);
+            _headerStaticIndex = HeaderStaticIndexUnset;
         }
 
         public void Dispose()
         {
             if (_stringOctets != null)
             {
-                Pool.Return(_stringOctets, true);
```

Comment:
Was there a particular reason for clearing the pooled arrays on return before?

Reply:
I don't know. Safety in case the pooled header bytes contained a secret?

Are there any other places we cleared pooled headers?

Reply:
> Are there any other places we cleared pooled headers?

Not that I'm aware of. Pipelines doesn't [even in the default case when it uses ArrayPool.Shared](https://github.com/dotnet/runtime/blob/5dc1b4c31312b6778b62ee28ba4a2b0c54f34898/src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/BufferSegment.cs#L73). In Kestrel's and IIS's case it's a separate `PinnedBlockMemoryPool` that isn't shared with other parts of the app, though the uncleared data is still accessible via the response `BodyWriter` if you go looking for it.

I wonder if using the `PinnedBlockMemoryPool` for this in Kestrel would be worthwhile. Maybe as a separate change.